pkg/query-service/app/querier/v2/helper.go

```diff
@@ -138,13 +210,36 @@ func (q *querier) runBuilderQuery(
 				}
 				return
 			}
+
+			filteredSeries, startTime, endTime := filterSeriesPoints(series, miss.Start, miss.End, builderQuery.StepInterval)
+
+			// making sure that empty range doesn't doesn't enter the cache
+			if len(filteredSeries) > 0 {
+				filteredMissedSeries = append(filteredMissedSeries, querycache.CachedSeriesData{
+					Data:  filteredSeries,
+					Start: startTime,
+					End:   endTime,
+				})
+			}
+
+			// for the actual response
 			missedSeries = append(missedSeries, querycache.CachedSeriesData{
 				Data:  series,
 				Start: miss.Start,
 				End:   miss.End,
 			})
 		}
-		mergedSeries := q.queryCache.MergeWithCachedSeriesData(cacheKeys[queryName], missedSeries)
+
+		filteredMergedSeries, err := q.queryCache.MergeWithCachedSeriesDataV2(cacheKeys[queryName], filteredMissedSeries)
+		if err != nil {
+			filteredMergedSeries = filteredMissedSeries
+		}
+		q.queryCache.StoreSeriesInCache(cacheKeys[queryName], filteredMergedSeries)
```

Comment:
What's the rationale for `MergeWithCachedSeriesDataV2` return err?

The error here is not the kind of error that the caller is expected to know and handle. As in if the calle doesn't know how to handle an error and should propagate up, then it makes sense to return an error to the caller so that the caller has control flow to handle it. In this case, error comes from the json.Unmarshal, which is the internal detail of the calle and it knows what to do in that case. Why do we need to prop that up and do the exact thing in caller.

Reply:
yeah make sense to fix the MergeWithCachedSeriesDataV2